web/src/peer_data.ts

```diff
@@ -1,17 +1,49 @@
+import {z} from "zod";
+
 import * as blueslip from "./blueslip.ts";
+import * as channel from "./channel.ts";
 import {LazySet} from "./lazy_set.ts";
+import {page_params} from "./page_params.ts";
 import type {User} from "./people.ts";
 import * as people from "./people.ts";
 import * as sub_store from "./sub_store.ts";
 
 // This maps a stream_id to a LazySet of user_ids who are subscribed.
 const stream_subscribers = new Map<number, LazySet>();
 const fetched_stream_ids = new Set<number>();
+export function has_full_subscriber_data(stream_id: number): boolean {
+    return fetched_stream_ids.has(stream_id);
+}
+const pending_subscriber_requests = new Map<number, Promise<LazySet>>();
 
 export function clear_for_testing(): void {
     stream_subscribers.clear();
 }
 
+const fetch_stream_subscribers_response_schema = z.object({
+    subscribers: z.array(z.number()),
+});
+
+export async function maybe_fetch_stream_subscribers(stream_id: number): Promise<LazySet> {
+    if (pending_subscriber_requests.has(stream_id)) {
+        return pending_subscriber_requests.get(stream_id)!;
+    }
+    const promise = new Promise<LazySet>((resolve) => {
+        channel.get({
+            url: "/json/streams/" + stream_id + "/members",
+            success(response) {
```

Comment:
How do we want to handle errors here? Maybe blueslip error and then resolve with whatever subscribers we already knew about?